docs/extensions/language-support.md

```diff
@@ -0,0 +1,861 @@
+classes and methods or language server protocol messages they map to.
+
+After that we'll list some methods that are available to manager the state in your extension.
+
+
+## Syntax Highlighting
+
+![syntax highlighting](./images/language-support/syntax-highlighting.png)
+
+In order to support syntax highlighting your extension needs to registers a TextMate grammar for its
+language in its `package.json` file.
+```json
+"contributes": {
+	"languages": [
+		{
+			"id": "markdown",
+			...
+		}
+	],
+	"grammars": [
+		{
+			"language": "markdown",
+			"scopeName": "text.html.markdown",
+			"path": "./syntaxes/markdown.tmLanguage"
+		}
+	], ...
+}
+```
+
+#### Basic
+ Start with a simple grammar that supports colorization of strings, comments, keywords, etc.
+
+#### Advanced
+Provide a grammar that understands terms and expressions and thus supports colorization of variables and function references etc.
+
+
+
+## Provide Out-Of-The-Box Code Snippets
+
+![Snippets at work](./images/language-support/snippets.gif)
+
+With snippets you can provide useful code template with placeholders. You need to register a file that contains the
+snippets for your language in your extension's `package.json` file.
+
+```json
+"snippets": [
+	{
+		"language": "javascript",
+		"path": "./snippets/javascript.json"
+	}, ...
+]
+```
+
+#### Basic
+Provide snippets with placeholders such as this example for `markdown`:
+
+```json
+"Insert ordered list": {
+	"prefix": "ordered list",
+	"body": [
+		"1. ${first}",
+		"2. ${second}",
+		"3. ${third}",
+		"$0"
+	],
+	"description": "Insert ordered list"
+}
+```
+
+#### Advanced
+Provide snippets that use explicit tab stops to guide the user and use nested placeholders such
+as this example for `groovy`
+
+```json
+"key: \"value\" (Hash Pair)": {
+		"prefix": "key",
+		"body": "${1:key}: ${2:\"${3:value}\"}"
+	}
+```
+
+
+
+## Support Smart Editing
+
+![Smart Editing At Work](./images/language-support/smart-editing.gif)
+
+You can provide a language configuration in your extension's `package.json` file.
+
+```json
+"contributes": {
+	"languages": [
+		{
+			"id": "typescript",
+			...
+			"configuration": "./language-configuration.json"
+		}, ...
+	]
+	...
+}
+```
+
+#### Basic
+None
+
+#### Advanced
+Here is an example from `TypeScript`:
+
+```json
+{
+	"comments": {
+		"lineComment": "//",
+		"blockComment": [ "/*", "*/" ]
+	},
+	"brackets": [
+		["{", "}"],
+		["[", "]"],
+		["(", ")"],
+		["<", ">"]
+	],
+	"autoClosingPairs": [
+		{ "open": "{", "close": "}" },
+		{ "open": "[", "close": "]" },
+		{ "open": "(", "close": ")" },
+		{ "open": "'", "close": "'", "notIn": ["string", "comment"] },
+		{ "open": "\"", "close": "\"", "notIn": ["string"] },
+		{ "open": "`", "close": "`", "notIn": ["string", "comment"] },
+		{ "open": "/**", "close": " */", "notIn": ["string"] }
+	],
+	"surroundingPairs": [
+		["{", "}"],
+		["[", "]"],
+		["(", ")"],
+		["<", ">"],
+		["'", "'"],
+		["\"", "\""],
+		["`", "`"]
+	]
+}
+```
+
+
```

Comment:
Add a header and paragraph saying this is now the language server section.  It would also be good to mention the registerBlahProvider pattern used that takes a language Id (GO_MODE in these examples).